### PR TITLE
IGNITE-13734 .NET: Register service return type on method invocation

### PR DESCRIPTION
**Pull request by nizhikov**
This PR fixes bug that occurs on the invocation of the service method that has not registered return type.

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by nizhikov** [comment]
Hello, @ptupitsyn 
Please, take a look at this fix.

Currently, we unable to read binary type from service response if it not registered explicitly.
This issue continues to fix transparent service invocation like #8568, [IGNITE-10075](https://issues.apache.org/jira/browse/IGNITE-10075).

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServiceProxyTest.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@
 namespace Apache.Ignite.Core.Tests.Services
 {
     using System;
+    using System.CodeDom;
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServiceProxyTest.cs:21]: Unused import

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs [hunk 2]
```diff
@@ -162,7 +168,34 @@ public BinaryType RegisterEnum(string typeName, IEnumerable<KeyValuePair<string,
         /// <returns>Type or null.</returns>
         public string GetTypeName(int id)
         {
-            return DoOutInOp((int) Op.GetType, w => w.WriteInt(id), r => Marshaller.StartUnmarshal(r).ReadString());
+            try
+            {
+                return DoOutInOp((int) Op.GetType, w =>
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs:173]: Let's extract a method `string GetTypeName(int id, bool java)`.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs [hunk 3]
```diff
@@ -162,7 +168,34 @@ public BinaryType RegisterEnum(string typeName, IEnumerable<KeyValuePair<string,
...
+            // Try to get java type name and register corresponding DotNet type.
+            var javaTypeName = DoOutInOp((int) Op.GetType, w =>
+            {
+                w.WriteInt(id);
+                w.WriteByte(JavaPlatformId);
+            }, r => Marshaller.StartUnmarshal(r).ReadString());
+
+            RegisterType(id, Marshaller.GetTypeName(javaTypeName), false);
+
+            return DoOutInOp((int) Op.GetType, w =>
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs:194]: I think we can return the name of the registered type from above, no need to call into Java again.

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs (hunk 2 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs:173]: Fixed.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs (hunk 3 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/BinaryProcessor.cs:194]: Fixed.

**Comment by nizhikov** [comment]
@ptupitsyn Thanks for the review.
All comments fixed.
Please, take a look one more time.

**Review by ptupitsyn — APPROVED** [review]

**Pull request merged**